File: Data Structures & Algorithms/matrixDFS/submission-4.py
class Solution:
    def countPaths(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        cols = len(grid[0])

        def dfs(r, c, visited):
            if min(r, c) < 0 or r == rows or c == cols or grid[r][c] == 1 or (r, c) in visited:
                return 0

            if r == rows - 1 and c == cols - 1:
                return 1

            visited.add((r, c))
            count = 0

            directions = [[-1, 0], [1, 0], [0, -1], [0, 1]]
            for dr, dc in directions:
                count += dfs(r + dr, c + dc, visited)
            
            visited.remove((r, c))
            return count

        return dfs(0, 0, set())

File: Data Structures & Algorithms/island-perimeter/submission-4.ts
class Solution {
    /**
     * @param {number[][]} grid
     * @return {number}
     */
    islandPerimeter(grid: number[][]): number {
        const ROWS = grid.length
        const COLS = grid[0].length
        const directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        const seen = new Set<string>()

        const dfs = (r: number, c: number) => {
            if (Math.min(r, c) < 0 || r >= ROWS || c >= COLS || grid[r][c] === 0) {
                return 1
            }

            if (seen.has(`#${r}#${c}`)) return 0

            seen.add(`#${r}#${c}`)

            let perimeter = 0
            for (const [dr, dc] of directions) {
                perimeter += dfs(r + dr, c + dc)
            }
            return perimeter
        }

        for (let r = 0; r < ROWS; r++) {
            for (let c = 0; c < COLS; c++) {
                if (grid[r][c] === 1) return dfs(r, c)
            }
        }
        return 0
    }
}
